Wikipedia/Code/DiffListViewController.swift

```diff
@@ -0,0 +1,322 @@
+    private let chunkedHeightCalculationsConcurrentQueue = DispatchQueue(label: "org.wikipedia.diff.chunkedHeightCalculations", qos: .userInteractive, attributes: .concurrent)
+    private let layoutSubviewsHeightCalculationsSerialQueue = DispatchQueue(label: "org.wikipedia.diff.layoutHeightCalculations", qos: .userInteractive)
+    
+    init(theme: Theme, delegate: DiffListDelegate?, type: DiffContainerViewModel.DiffType) {
+        self.type = type
+        super.init(nibName: nil, bundle: nil)
+        self.theme = theme
+        self.delegate = delegate
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        view.addSubview(collectionView)
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        view.addConstraints([
+            view.safeAreaLayoutGuide.leadingAnchor.constraint(equalTo: collectionView.leadingAnchor),
+            view.safeAreaLayoutGuide.trailingAnchor.constraint(equalTo: collectionView.trailingAnchor),
+            view.topAnchor.constraint(equalTo: collectionView.topAnchor),
+            view.bottomAnchor.constraint(equalTo: collectionView.bottomAnchor)
+        ])
+        collectionView.register(DiffListChangeCell.wmf_classNib(), forCellWithReuseIdentifier: DiffListChangeCell.reuseIdentifier)
+        collectionView.register(DiffListContextCell.wmf_classNib(), forCellWithReuseIdentifier: DiffListContextCell.reuseIdentifier)
+        collectionView.register(DiffListUneditedCell.wmf_classNib(), forCellWithReuseIdentifier: DiffListUneditedCell.reuseIdentifier)
+    }
+    
+        override func viewDidLayoutSubviews() {
+            super.viewDidLayoutSubviews()
+            
+            if updateWidthsOnLayoutSubviews {
+                
+                //possibly overkill...simple main thread way to do this is updateListViewModels(listViewModel: self.dataSource, updateType: updateType) followed by the main thread lines 78-83. This does seem very slightly faster for rotation though, and could yield more gains as diffs get larger. More improvements could be size caching & putting layoutSubviewsHeightCalculationsSerialQueue instead into an NSOperation to be cancelled if another viewDidLayoutSubviews is called. No need to continue calculating sizes if collection view frame has changed again.
+                //todo: clean up - move this and updateListViewModel methods into separate class, DiffListSizeCalculator or something
+                let updateType = ListUpdateType.layoutUpdate(collectionViewWidth: self.collectionView.frame.width, traitCollection: self.traitCollection)
+                
+                //actually not sure if this serial queue is needed or simply calling on the main thread (also serial) is the same. this also seems faster than without though.
+                layoutSubviewsHeightCalculationsSerialQueue.async {
```

Comment:
I thought queues with qos set to interactive `userInteractive` would run the tasks on the main thread but that might not be true according to https://stackoverflow.com/a/49658842/4574147

Reply:
Yeah, I definitely see the code within both queues running on non-main threads. The main thread checker never fires (this resizes all the heights upon rotation) so I guess it's happy. I figured `userInteractive` fit better here than `userInitiated` since it needs to happen during a rotation, but I haven't measured the difference between the two.